### PR TITLE
Added single pipe flow process based on the HeatTransportBHE

### PR DESCRIPTION
**Pull request by ChaofanChen**
1. The single pipe flow process was implemented based on the HeatTransportBHE process, in which the fluid flows through the pipe buried in the wellbore and exchanging heat energy with surrounding formation.
2. New BHE boundary condition that coupled with heat pump representing by the COP curve was added. 
3. Tests covering the single pipe flow were added.
4. New feature and behavior change were documented including 1P process and new BHE boundary condition.

### CODE REVIEW

File: ProcessLib/HeatTransportBHE/BHE/BHE_1P.cpp [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    return {{
+        // pipe, Eq. 19
+        (lambda_r + rho_r * Cp_r * alpha_L * _flow_velocity),
+        // grout, Eq. 21
+        (1.0 - porosity_g) * lambda_g,
+    }};
+}
+
+std::array<Eigen::Vector3d, BHE_1P::number_of_unknowns>
+BHE_1P::pipeAdvectionVectors(Eigen::Vector3d elem_direction_vec) const
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/BHE_1P.cpp]: Should be a const reference, unless the current way is faster.
```suggestion
BHE_1P::pipeAdvectionVectors(Eigen::Vector3d const& element_direction) const
```
Also no need to repeat the type in argument's name.

File: ProcessLib/HeatTransportBHE/BHE/BHE_1P.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+{
+    auto const tm_flow_properties = calculateThermoMechanicalFlowPropertiesPipe(
+        _pipe.single_pipe, borehole_geometry.length, refrigerant, flow_rate);
+
+    _flow_velocity = tm_flow_properties.velocity;
+    _thermal_resistances =
+        calcThermalResistances(tm_flow_properties.nusselt_number);
+}
+
+/// Nu is the Nusselt number.
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/BHE_1P.cpp]:
```suggestion
// Nu is the Nusselt number.
```
Otherwise it is a documentation for the function.
You could do it, in the header though, with `\param Nu ...`

File: ProcessLib/HeatTransportBHE/BHE/BuildingPowerCurves.h [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ */
+
+#pragma once
+
+#include <map>
+#include <memory>
+#include <string>
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/BuildingPowerCurves.h]:
```suggestion
```
It would be nice, if you check other includes too.

File: ProcessLib/HeatTransportBHE/BHE/BuildingPowerCurves.h [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+#include "BaseLib/ConfigTree.h"
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/BuildingPowerCurves.h]:
```suggestion
```

File: ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.cpp [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+ * \copyright
+ * Copyright (c) 2012-2020, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ */
+
+#include "BaseLib/ConfigTree.h"
+#include "CreateBHEUType.h"
+
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.cpp]: The map, memory etc includes are probably needed here...

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp [hunk 6]
```diff
@@ -283,14 +283,37 @@ void HeatTransportBHEProcess::createBHEBoundaryConditionTopBottom(
...
+
+            int cnt = 0;
+
+            for (auto const element : connected_elems)
+            {
+                if (element->getDimension() == 1)
+                {
+                    cnt++;
+                }
+            }
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: std::count_if

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp [hunk 7]
```diff
@@ -303,11 +326,33 @@ void HeatTransportBHEProcess::createBHEBoundaryConditionTopBottom(
...
+                  _process_data._vec_BHE_property[bhe_i]);
 
         auto createBCs = [&](auto& bhe) {
             for (auto const& in_out_component_id :
                  bhe.inflow_outflow_bc_component_ids)
             {
+                // Top, inflow, normal case
+                // The single pipe (1P type) flow process has 2 primary
+                // variables. Its inflow and outflow nodes are different.
+                static constexpr int number_of_unknowns_for_1P_type = 2;
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: The FEM part should be independent of the BHE-type.

File: ProcessLib/HeatTransportBHE/LocalAssemblers/HeatTransportBHELocalAssemblerBHE.h [hunk 8]
```diff
@@ -29,14 +29,13 @@ class HeatTransportBHELocalAssemblerBHE
     : public HeatTransportBHELocalAssemblerInterface
 {
     static constexpr int bhe_unknowns = BHEType::number_of_unknowns;
-    static constexpr int single_bhe_unknowns_size = ShapeFunction::NPOINTS;
-    static constexpr int temperature_size = ShapeFunction::NPOINTS;
-    static constexpr int temperature_index = 0;
-    static constexpr int bhe_unknowns_size =
-        single_bhe_unknowns_size * bhe_unknowns;
+    static constexpr int n_int_points = ShapeFunction::NPOINTS;
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/LocalAssemblers/HeatTransportBHELocalAssemblerBHE.h]: I'm probably missing something, but the number of integration points is not equal to the number of shape functions in general.

File: ProcessLib/HeatTransportBHE/LocalAssemblers/HeatTransportBHELocalAssemblerBHE-impl.h [hunk 9]
```diff
@@ -57,20 +57,27 @@ HeatTransportBHELocalAssemblerBHE<ShapeFunction, IntegrationMethod, BHEType>::
         _secondary_data.N[ip] = sm.N;
     }
 
+    // calculate the element direction vector
+    const double* coords_1 = e.getNode(0)->getCoords();
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/LocalAssemblers/HeatTransportBHELocalAssemblerBHE-impl.h]:
```c++
p0 = Eigen::Map<Eigen::Vector3d>(...getCoords(), 3);
```
would be nicer.

File: ProcessLib/HeatTransportBHE/BHE/BHE_1P.cpp (hunk 2 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/BHE/BHE_1P.cpp]: Thanks, changed.

File: ProcessLib/HeatTransportBHE/BHE/BuildingPowerCurves.h (hunk 3 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/BHE/BuildingPowerCurves.h]: Sorry about that, deleted.

File: ProcessLib/HeatTransportBHE/BHE/BuildingPowerCurves.h (hunk 4 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/BHE/BuildingPowerCurves.h]: deleted.

File: ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.cpp (hunk 5 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.cpp]: Added in the header file.

File: ProcessLib/HeatTransportBHE/BHE/BHE_1P.cpp (hunk 1 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/BHE/BHE_1P.cpp]: changed.

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp (hunk 6 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: Thanks, that's better.

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp [hunk 10]
```diff
@@ -288,18 +288,14 @@ void HeatTransportBHEProcess::createBHEBoundaryConditionTopBottom(
...
-            for (auto const element : connected_elems)
-            {
-                if (element->getDimension() == 1)
-                {
-                    cnt++;
-                }
-            }
-            if (cnt == 1)
+            // Count number of 1d elements in every BHE node.
+            const std::size_t nElems = std::count_if(
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]:
```suggestion
            const std::size_t n_elements = std::count_if(
```
At least the camel-case should be fixed.

File: ProcessLib/HeatTransportBHE/BHE/BHE_1P.h [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+ * the grout zone sorrounding the single pipe T_g. These two primary varaibles
+ * are solved according to heat convection and conduction equations on the pipes
+ * and also in the grout zones. The interaction of the 1P type of BHE and the
+ * sorrounding soil is regulated through the thermal resistance values, which
+ * are calculated specifically during the initialization of the class.
+ */
+class BHE_1P final : public BHECommon
+{
+public:
+    BHE_1P(BoreholeGeometry const& borehole,
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/BHE_1P.h]: Most, if not all of the implementations could go into the cpp.

File: ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.h [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+
+#pragma once
+
+#include <map>
+#include <memory>
+#include <string>
+
+#include "BHECommon.h"
+#include "FlowAndTemperatureControl.h"
+#include "PipeConfiguration1PType.h"
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.h]: Are these needed here?

File: ProcessLib/HeatTransportBHE/BHE/PipeConfiguration1PType.h [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+ *
+ * \copyright
+ * Copyright (c) 2012-2020, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ */
+
+#pragma once
+#include "BaseLib/ConfigTree.h"
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/PipeConfiguration1PType.h]: dto.

File: ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp (hunk 10 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/HeatTransportBHEProcess.cpp]: Changed.

File: ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.h (hunk 12 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.h]: Don't need, deleted.

File: ProcessLib/HeatTransportBHE/BHE/PipeConfiguration1PType.h (hunk 13 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/BHE/PipeConfiguration1PType.h]: Deleted.

File: ProcessLib/HeatTransportBHE/BHE/BHE_1P.h [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ */
+
+#pragma once
+
+#include <Eigen/Eigen>
+
+#include "BaseLib/Error.h"
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/BHE_1P.h]: This include isn't necessary, isn't it?

File: ProcessLib/HeatTransportBHE/BHE/BuildingPowerCurves.h (hunk 3 above)
Comment (endJunction) [reply ProcessLib/HeatTransportBHE/BHE/BuildingPowerCurves.h]: > It would be nice, if you check other includes too.
I mean all other includes in all other files too, not just here.

Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/BHE/BuildingPowerCurves.h]: Got it.

File: ProcessLib/HeatTransportBHE/LocalAssemblers/HeatTransportBHELocalAssemblerBHE.h (hunk 8 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/LocalAssemblers/HeatTransportBHELocalAssemblerBHE.h]: Yes, but in BHE elements, which are line elements only. And there is no need to use higher order shape function considering simulation purpose. So the intergration point here in BHE line elements should be 2. For better readability, we changed the `unknowns_size` to the `n_int_points`. If you insist, I can also remove this change.

Comment (endJunction) [reply ProcessLib/HeatTransportBHE/LocalAssemblers/HeatTransportBHELocalAssemblerBHE.h]: The number of integration points is independent of the element type.
[integration_order](https://doxygen.opengeosys.org/ogs_file_param__prj__processes__process__integration_order.html)
[shape function order](https://doxygen.opengeosys.org/ogs_file_param__prj__process_variables__process_variable__order.html)
Both can be set through the project file. So you can have linear line elements with fourth integration order. This is used *e.g.* in richards-mechanics, where there is high non-linearity in certain setups.

File: ProcessLib/HeatTransportBHE/LocalAssemblers/HeatTransportBHELocalAssemblerBHE-impl.h (hunk 9 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/LocalAssemblers/HeatTransportBHELocalAssemblerBHE-impl.h]: Thanks, changed.

File: ProcessLib/HeatTransportBHE/BHE/BHE_1P.h (hunk 11 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/BHE/BHE_1P.h]: Moved the constructor and calculation of cross section areas into cpp file.

**Comment by endJunction** [comment]
@ChaofanChen Could you rebase the PR to current master. In the master some issues were fixed, so more of the CI will be green.

I'm looking in the BHE/FEM separation issue now.

**Comment by endJunction** [comment]
So far all urgent issues were resolved. Needs a rebase, ask if you need help, and optional commits cleanup.
@renchao-lu Could you do second review after the rebase?

File: ProcessLib/HeatTransportBHE/BHE/BHECommonCoaxial.cpp [hunk 15]
```diff
@@ -130,6 +131,25 @@ BHECommonCoaxial::calcThermalResistances(double const Nu_inner_pipe,
     return getThermalResistances(R_gs, R_ff, R_fg);
 }
 
+std::array<std::pair<std::size_t /*node_id*/, int /*component*/>, 2>
+BHECommonCoaxial::getBHEInflowDirichletBCNodesAndComponents(
+    std::size_t top_node_id,
+    std::size_t /*bottom_node_id*/,
+    int in_component_id)
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/BHECommonCoaxial.cpp]: For safety, this arguments and the function should be `const`.

File: ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.cpp [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+    BaseLib::ConfigTree const& config,
+    std::map<std::string,
+             std::unique_ptr<MathLib::PiecewiseLinearInterpolation>> const&
+        curves)
+{
+    // if the BHE is using python boundary condition
+    auto const bhe_if_use_python_bc_conf =
+        config.getConfigParameter<bool>("use_bhe_pipe_network", false);
+    DBUG("If using python boundary condition : %s",
+         (bhe_if_use_python_bc_conf) ? "true" : "false");
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.cpp]: Such messages should be statements.
```suggestion
    if (use_python_bcs) { DBUG("BHE 1P using python boundary conditions."; }
```
Please note the change in the variable name.

File: ProcessLib/HeatTransportBHE/BHE/CreateFlowAndTemperatureControl.cpp [hunk 17]
```diff
@@ -86,6 +87,31 @@ FlowAndTemperatureControl createFlowAndTemperatureControl(
                                       refrigerant.specific_heat_capacity,
                                       refrigerant.density};
     }
+
+    if (type == "BuildingPowerCurveConstantFlow")
```
Comment (endJunction) [inline ProcessLib/HeatTransportBHE/BHE/CreateFlowAndTemperatureControl.cpp]: Either the class name should be `BuildingPowerCurveConstantFlow` or the type here should be `BuildingPowerCurve`.

**Comment by endJunction** [comment]
Not sure if you want to keep the commits like they are, or going to clean it up before merging.
Add the `const`, then we can merge I think.

Doxygen docu needs some love too: https://jenkins.opengeosys.org/job/ufz/job/ogs/job/PR-2818/19/Doxygen/  QA page for parameters.

File: ProcessLib/HeatTransportBHE/BHE/BHECommonCoaxial.cpp (hunk 15 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/BHE/BHECommonCoaxial.cpp]: Added.

File: ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.cpp (hunk 16 above)
Comment (ChaofanChen) [reply ProcessLib/HeatTransportBHE/BHE/CreateBHE1PType.cpp]: Changed.

**Comment by ChaofanChen** [comment]
@endJunction The current commits are good to me, please let me know if you have any suggestions on cleaning them.
